File: 61. Rotate List-2.cpp
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode *rotateRight(ListNode *head, int k) {
        if (!head) return nullptr;
        int n = 1;
        // find end and count num of nodes
        ListNode * curr = head;
        while(curr->next){
            n++;
            curr = curr->next;
        }
        ListNode * end = curr;
        end->next = head;

        // find where to truncate
        curr = end;
        for(int i = 0; i < n-k;i++){
            curr = curr->next;
        }
        ListNode* newHead = curr->next;
        curr->next=nullptr;
        return newHead;
    }
};

File: 61. Rotate List-2.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
        
from typing import Optional
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        # find end and count num of nodes
        if not head:
            return None
        curr = head
        n = 1
        while curr.next:
            n += 1
            curr = curr.next
        end = curr
        end.next = head
        
        # find where to truncate
        curr = end
        for i in range(n-k):
            curr = curr.next
        newHead = curr.next
        curr.next = None
        
        return newHead
        
        
        
            
            